# This is a Jupyter (Python) Notebook

### A Jupyter Notebook is the most common way scientists organize their programs.  
### It's attractive because it organizes text, code, figures, into a single document. 
### Its not a **program** in the conventional sense of the term, in that it is usually executed in an interactive way. 
### All of the lectures, problem sets, tests, homeworks, in this class will be provided in the form of a Jupyter notebook. 


# Visual Studio Code is an IDE

### A Jupyter notebook can be opened in different Interactive Development Environments (IDE).  
### My favorite is Visual Studio Code (VS Code) which I use for my research.  
### Another widely used IDE is **Jupyter Lab** /**Jupyter Notebook**.  
### They are kind of the same thing, and Jupyter Lab is the new and improved Jupyter Notebook, but not all features have made it over. 
### If you already took COGS/PSYC 10A and really loved Rstudio, or are familiar with Matlab you could use **Spyder** which has the same look and feel.  In Spyder you need to install the Jupyter Notebook extension.  

## Cells

### This is a cell, which is indicated by the bar on the left. 
### Cells are of different types.  This is a **markdown** cell that contains text. 
### A markdown cell contains html code, and is where i was place Instructional text in the Lectures and Tutorials, and the Problem Statement in Homework and Tests.
### If you double click on it it can be edited.  After you are done, click on the checkbox to render it (or SHIFT-ENTER)   
### The next cell is labeled "Python" and contains a bit of python code you can run. 
### Click on the next cell, then click on the arrow to the left of the box. Alternatively, you can click type shift-enter to execute the code. 

In [1]:
#notice that after I execute the cell, the number 1 appears in the box on the left.  IF I execute it again, the number changes to 2. 
variable_name = "my_variable"
print(variable_name)
#notice that the output of the cell appears below the cell box.  This will be true for print statement and for figures. It will also be where error messages can appear. 

my_variable


## Kernel 
### Its important that you track the kernel that you are using for Python.  
### On the top right, is an icon that either says "Select Kernel" or it says "base(Python 3.11)"  
### If it says "Select Kernel" you need to choose one before executing.  
### if it does not say "base" or "conda" or "anaconda" you need to change the kernel. Notice, there is a kernel that comes with your system. You do not want to use this kernel.  

## The `print` statement
### The next cell does a small calculation of adding 2 numbers.  The print command displays the output

In [2]:
a = 1
b = 2

c = a + b

print(c)

3


### The next cell does not print. Try running it.  It will seem to do nothing.
### You need a `print` command to see the output. 
### Notice I used a `#` symbol to indicate a comment and not code. 
### I wrote a comment inline to indicate what the operation was on the line that I multiplied. 
### On the next line I put a comment symbol `#` in front of print(c).  This means this line is not being executed. 
### Next try removing the comment and running the cell again.  

In [4]:
a = 5 
b = 4

c = a*b  #multiply


## The **VARIABLES** or **JUPYTER** window
### Another way to examine the output is to directly look at the variables in a table.
### The Variables Tab will open a window with your variables. 
### By default, the variables will appear below your python.  If you want to move it to the right
### View -> Command Pallette -> View:Focus into Secondary Side Bar
### Then drag the variables pane to the right. 

### In the next example I do a calculation but I do not set it equal to any variable.  Try running it. 

In [5]:
a = 5
b = 4
a/b #divide

1.25

### When there was nowhere to put the output of the divide, it just output to the window
### Take a look at the VARIABLES table.  There is no variable with value 1.25
### That value is lost forever!

### Let's create an empty cell and make a new example with 
### a = 3 
### b = 2 
### and compute c = a-b and display the results.
  

In [6]:
a = 3
b = 2 
c = a -b 
print('c =', c)

c = 1


### Remember that the Cell Type should be **Python** in order to execute the cell. 
### A cell that has text material is labeled **Markdown**.
### A cell that has programs is labeled **Code**  
### There are two ways to add a cell. One is in the bar on the top, where there is a `+` button that will add the cell of each type. 
### The other way is the rectangular box below this cell. This can inseert a cell below.

## SAVING YOUR WORK. 
### Visual Studio Code will autosave as you are going to temporary files.  But you need to save your work to the original file.  To do that, 
### please use File-> Save from the menu. 
## REVERTING YOUR WORK
### Suppose its all gone horribly wrong and you just want to restart everything.  
### You can reset the outputs of the jupyter notebook using the "Clear All Outputs" menu option under edit. 
### You can also restart the kernel using the circle button.  
### But *most usefully* you can revert to return it to your last good save. 
### just close the window at the filename, and do not save your work.  
### Under the file menu, reload the notebook from disk.  The latter option **removes all your work** since the last save. 


# Create a Class Folder in your Computer
### Setup a class folder in a location that you know on your computer, and keep all your files organized there.
### Use the Open Folder option in the file menu to point Visual Studio Code to that folder! 
### Visual Studio will remember where that folder is when you start it again.   